File: C#/TheOpenMathLibrary.ActuarialCalculators/PiecewiseSpecialFunctionsLibrary.cs
namespace TheOpenMathLibrary.ActuarialCalculators
{
    /// <summary>
    /// Provides piecewise-defined and special mathematical functions.
    /// </summary>
    public class PiecewiseSpecialFunctionsLibrary
    {
        private const double ComparisonTolerance = 1e-12;

        /// <summary>
        /// Evaluates the indicator function for zero.
        /// </summary>
        /// <param name="x">The value to test.</param>
        /// <returns><c>1</c> when <paramref name="x"/> is zero; otherwise <c>0</c>.</returns>
        public static double IndicatorFunction(double x)
        {
            return x == 0d ? 1d : 0d;
        }

        /// <summary>
        /// Evaluates the unit step function with a midpoint value of 0.5 at zero.
        /// </summary>
        /// <param name="x">The value at which to evaluate the function.</param>
        /// <returns>The step-function value at <paramref name="x"/>.</returns>
        public static double StepFunction(double x)
        {
            if (x < 0d)
            {
                return 0d;
            }

            return x == 0d ? 0.5d : 1d;
        }

        /// <summary>
        /// Evaluates the Heaviside step function.
        /// </summary>
        /// <param name="x">The value at which to evaluate the function.</param>
        /// <returns><c>0</c> for negative values; otherwise <c>1</c>.</returns>
        public static double HeavisideStepFunction(double x)
        {
            return x < 0d ? 0d : 1d;
        }

        /// <summary>
        /// Evaluates the rectangular function centered at zero.
        /// </summary>
        /// <param name="x">The value at which to evaluate the function.</param>
        /// <returns>The rectangular-function value at <paramref name="x"/>.</returns>
        public static double RectangularFunction(double x)
        {
            if (x < -0.5d || x > 0.5d)
            {
                return 0d;
            }

            return Math.Abs(x - 0.5d) <= ComparisonTolerance || Math.Abs(x + 0.5d) <= ComparisonTolerance ? 0.5d : 1d;
        }

        /// <summary>
        /// Evaluates a sawtooth wave with period 1.
        /// </summary>
        /// <param name="x">The input value.</param>
        /// <returns>The sawtooth-wave value at <paramref name="x"/>.</returns>
        public static double SawtoothFunction(double x)
        {
            return x - Math.Floor(x);
        }

        /// <summary>
        /// Evaluates a triangle wave with period 1 and range [0, 1].
        /// </summary>
        /// <param name="x">The input value.</param>
        /// <returns>The triangle-wave value at <paramref name="x"/>.</returns>
        public static double TriangleWaveFunction(double x)
        {
            return 2d * Math.Abs(SawtoothFunction(x) - 0.5d);
        }

        /// <summary>
        /// Evaluates a square wave derived from the sawtooth function.
        /// </summary>
        /// <param name="x">The input value.</param>
        /// <returns><c>0</c> on the first half-period and <c>1</c> on the second half-period.</returns>
        public static double SquareWaveFunction(double x)
        {
            return HeavisideStepFunction(SawtoothFunction(x) - 0.5d);
        }

        /// <summary>
        /// Evaluates the normalized sinc function sin(x) / x.
        /// </summary>
        /// <param name="x">The input value.</param>
        /// <returns>The sinc value at <paramref name="x"/>.</returns>
        public static double SincFunction(double x)
        {
            if (Math.Abs(x) <= ComparisonTolerance)
            {
                return 1d;
            }

            return Math.Sin(x) / x;
        }

        /// <summary>
        /// Evaluates the Dirichlet kernel of order <paramref name="order"/>.
        /// </summary>
        /// <param name="x">The input angle in radians.</param>
        /// <param name="order">The non-negative kernel order.</param>
        /// <returns>The value of the Dirichlet kernel.</returns>
        /// <exception cref="ArgumentOutOfRangeException">Thrown when <paramref name="order"/> is negative.</exception>
        public static double DirichletKernel(double x, int order)
        {
            if (order < 0)
            {
                throw new ArgumentOutOfRangeException(nameof(order), "The order must be non-negative.");
            }

            if (Math.Abs(x) <= ComparisonTolerance)
            {
                return 2d * order + 1d;
            }

            return Math.Sin((order + 0.5d) * x) / Math.Sin(0.5d * x);
        }
    }
}

File: Rust/src/actuarial_and_arithmetic/piecewise_special_functions_library.rs
//! Piecewise-defined elementary and waveform-style special functions.

/// Provides simple piecewise and periodic helper functions.
pub struct PiecewiseSpecialFunctionsLibrary;

impl PiecewiseSpecialFunctionsLibrary {
    const EPSILON: f64 = 1e-12;

    fn is_close(value_a: f64, value_b: f64) -> bool {
        (value_a - value_b).abs() <= Self::EPSILON
    }

    fn phase(x: f64) -> f64 {
        x - x.floor()
    }

    /// Returns `1.0` when `x` is approximately zero and `0.0` otherwise.
    pub fn indicator_function(x: f64) -> f64 {
        if Self::is_close(x, 0.0) {
            1.0
        } else {
            0.0
        }
    }

    /// Returns the sign of `x` using a small tolerance around zero.
    pub fn sign_function(x: f64) -> f64 {
        if x > Self::EPSILON {
            1.0
        } else if x < -Self::EPSILON {
            -1.0
        } else {
            0.0
        }
    }

    /// Returns the symmetric step function value at `x`.
    pub fn step_function(x: f64) -> f64 {
        if x < -Self::EPSILON {
            0.0
        } else if Self::is_close(x, 0.0) {
            0.5
        } else {
            1.0
        }
    }

    /// Returns the Heaviside step function value at `x`.
    pub fn heaviside_step_function(x: f64) -> f64 {
        if x < 0.0 {
            0.0
        } else {
            1.0
        }
    }

    /// Returns the rectangular pulse value at `x`.
    pub fn rectangular_function(x: f64) -> f64 {
        if !(-0.5 - Self::EPSILON..=0.5 + Self::EPSILON).contains(&x) {
            0.0
        } else if Self::is_close(x, -0.5) || Self::is_close(x, 0.5) {
            0.5
        } else {
            1.0
        }
    }

    /// Returns the sawtooth wave value at `x` with period `1`.
    pub fn sawtooth_function(x: f64) -> f64 {
        Self::phase(x)
    }

    /// Returns the triangle wave value at `x` with period `1`.
    pub fn triangle_wave_function(x: f64) -> f64 {
        1.0 - 4.0 * (Self::phase(x) - 0.5).abs()
    }

    /// Returns the square wave value at `x` with period `1`.
    pub fn square_wave_function(x: f64) -> f64 {
        let phase = Self::phase(x);

        if Self::is_close(phase, 0.0) || Self::is_close(phase, 0.5) {
            0.0
        } else if phase < 0.5 {
            1.0
        } else {
            -1.0
        }
    }

    /// Returns the normalized sinc function value at `x`.
    pub fn sinc_function(x: f64) -> f64 {
        if Self::is_close(x, 0.0) {
            1.0
        } else {
            x.sin() / x
        }
    }

    /// Returns the Dirichlet kernel of order `n` evaluated at `x`.
    pub fn dirichlet_kernel(x: f64, n: u32) -> f64 {
        let half_x = 0.5 * x;

        if half_x.sin().abs() <= Self::EPSILON {
            (2 * n + 1) as f64
        } else {
            (((n as f64) + 0.5) * x).sin() / half_x.sin()
        }
    }
}

#[cfg(test)]
mod tests {
    use super::PiecewiseSpecialFunctionsLibrary;

    #[test]
    fn elementary_piecewise_functions_match_expected_values() {
        assert_eq!(
            PiecewiseSpecialFunctionsLibrary::indicator_function(0.0),
            1.0
        );
        assert_eq!(PiecewiseSpecialFunctionsLibrary::sign_function(-3.0), -1.0);
        assert_eq!(PiecewiseSpecialFunctionsLibrary::step_function(0.0), 0.5);
        assert_eq!(
            PiecewiseSpecialFunctionsLibrary::heaviside_step_function(-1.0),
            0.0
        );
        assert_eq!(
            PiecewiseSpecialFunctionsLibrary::rectangular_function(0.25),
            1.0
        );
    }

    #[test]
    fn periodic_waves_are_shaped_correctly() {
        assert!((PiecewiseSpecialFunctionsLibrary::sawtooth_function(1.75) - 0.75).abs() < 1e-12);
        assert!(
            (PiecewiseSpecialFunctionsLibrary::triangle_wave_function(0.25) - 0.0).abs() < 1e-12
        );
        assert_eq!(
            PiecewiseSpecialFunctionsLibrary::square_wave_function(0.25),
            1.0
        );
        assert_eq!(
            PiecewiseSpecialFunctionsLibrary::square_wave_function(0.75),
            -1.0
        );
    }

    #[test]
    fn special_functions_handle_removable_singularities() {
        assert!((PiecewiseSpecialFunctionsLibrary::sinc_function(0.0) - 1.0).abs() < 1e-12);
        assert!((PiecewiseSpecialFunctionsLibrary::dirichlet_kernel(0.0, 3) - 7.0).abs() < 1e-12);
    }
}
